packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1233,12 +1234,39 @@ function pushEffect(tag, create, destroy, deps) {
 
 function mountRef<T>(initialValue: T): {|current: T|} {
   const hook = mountWorkInProgressHook();
-  const ref = {current: initialValue};
   if (__DEV__) {
+    const fiber = currentlyRenderingFiber;
+    let current = initialValue;
+    let shouldWarnAboutRead = false;
+    const ref = {
+      get current() {
+        if (shouldWarnAboutRead && currentlyRenderingFiber !== null) {
+          console.warn(
+            '%s: Unsafe read of a mutable value during render.\n\n' +
+              'Reading from a ref during render is only safe if:\n' +
+              '1. The ref value has not been updated, or\n' +
+              '2. The ref holds a lazily-initialized value that is only set once.\n\n%s',
+            getComponentName(fiber.type) || 'Unknown',
+            getStackByFiberInDevAndProd(fiber),
+          );
+        }
+        return current;
+      },
+      set current(value) {
+        if (!shouldWarnAboutRead && current !== null) {
```

Comment:
Would this be valid if `current === undefined` (not `null`)? I've seen `useRef()` without args, even though `mountRef` requires an argument, it can be of void type.

Should it warn on writes within render but not within effect if `current !== initialValue` and `value !== initialValue`?

Reply:
> Should it warn on writes within render but not within effect if `current !== initialValue` and `value !== initialValue`?

I assume you mean reads, but...not. It *is* okay to read from the ref if it's changed from an initial `null` value (for the lazily initialization pattern) but *only if it's changed once*. If the initial value isn't `null` though, reading after any change could be problematic.

> Would this be valid if `current === undefined` (not `null`)? I've seen `useRef()` without args, even though `mountRef` requires an argument, it can be of void type.

Adding a case for `undefined` is probably reasonable.

Reply:
No, I meant **writes**.